File: taewan/08.16/49.cpp
#include<iostream>
#include<vector>
#include<string>
#include<map>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> answer;
        vector<string> temp = strs;
        map<string, vector<string>> m;
        for(int i = 0; i < temp.size(); i++)        //각각의 단어를 알파벳 순으로 정렬   eat => aet
            sort(temp[i].begin(), temp[i].end());   
        for(int i = 0; i < temp.size(); i++) {
            if(m.find(temp[i]) == m.end()) {        //map에 키값이 없다면 strs배열 값을 추가해준다.
                vector<string> vec;
                vec.push_back(strs[i]);
                m[temp[i]] = vec;
            } else m[temp[i]].push_back(strs[i]); //이미 있는 키값이라면 push
        }
        for(auto iter : m) answer.push_back(iter.second);            
        return answer;
    }
};

int main(void){
  return 0;
}

File: deprecated/hyeyoung/49.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        word_criterion = []
        answer = {}

        for word in strs:
            sort_word = sorted(list(word))  # 단어의 철자를 오름차순으로 정렬한다
            if sort_word not in word_criterion:  # 고유의 단어순과 일치되는게 없으면 answer에 키값으로 등록하고 단어 표준 (word_criterion)에 등록
                word_criterion.append(sort_word)
                answer[tuple(sort_word)] = [word]
            else:
                answer[tuple(sort_word)] += [word]
        result = answer.values()
        return result